File: src/backend/Backend.cc
#include "Backend.h"
#include <string>
#include <napi.h>

Backend::Backend(std::string name, Napi::CallbackInfo& info) : name(name), env(info.Env()) {
  int width  = 0;
  int height = 0;
  if (info[0].IsNumber()) width  = info[0].As<Napi::Number>().Int32Value();
  if (info[1].IsNumber()) height = info[1].As<Napi::Number>().Int32Value();
  this->width = width;
  this->height = height;
}

void Backend::setCanvas(Canvas* _canvas)
{
  this->canvas = _canvas;
}



std::string Backend::getName()
{
  return name;
}

int Backend::getWidth()
{
  return this->width;
}
void Backend::setWidth(int width_)
{
  this->destroySurface();
  this->width = width_;
}

int Backend::getHeight()
{
  return this->height;
}
void Backend::setHeight(int height_)
{
  this->destroySurface();
  this->height = height_;
}

bool Backend::isSurfaceValid() {
  bool isValid = true;

  cairo_status_t status = cairo_surface_status(ensureSurface());

  if (status != CAIRO_STATUS_SUCCESS) {
    error = cairo_status_to_string(status);
    isValid = false;
  }

  return isValid;
}


BackendOperationNotAvailable::BackendOperationNotAvailable(Backend* backend,
  std::string operation_name)
  : operation_name(operation_name)
{
  msg = "operation " + operation_name +
    " not supported by backend " + backend->getName();
};

BackendOperationNotAvailable::~BackendOperationNotAvailable() throw() {};

const char* BackendOperationNotAvailable::what() const throw()
{
  return msg.c_str();
};

File: src/backends/backend.rs
use cairo::{ffi::cairo_surface_destroy, Format, Surface};
use napi::Env;

use crate::{canvas::canvas::Canvas, impl_backend_base, impl_backend_trait};

#[derive(Debug, Clone, Default)]
pub struct BaseBackend {
    name: String,
    error: Option<String>,

    pub(crate) width: i32,
    pub(crate) height: i32,
    pub(crate) surface: Option<Surface>,
    pub(crate) canvas: Canvas<Box<BaseBackend>>,
}

pub trait BackendTrait: std::fmt::Debug + Clone + Sized {
    unsafe fn create_surface(&mut self, env: &mut Env) -> Option<Surface>;
    unsafe fn recreate_surface(&mut self, env: &mut Env) -> Option<Surface>;
    unsafe fn destroy_surface(&mut self, env: &mut Env);
    unsafe fn is_surface_valid(&mut self, env: &mut Env) -> bool;

    unsafe fn set_canvas(&mut self, canvas: Canvas);
    unsafe fn set_width(&mut self, width: i32, env: &mut Env);
    unsafe fn set_height(&mut self, height: i32, env: &mut Env);

    fn set_format(&mut self, format: Format);
    fn set_error(&mut self, error: String);

    fn get_surface(&self) -> Option<Surface>;
    fn get_name(&self) -> String;
    fn get_width(&self) -> i32;
    fn get_height(&self) -> i32;
    fn get_format(&self) -> Format;
    fn get_error(&self) -> Option<String>;
}

impl_backend_base!(BaseBackend);

impl BackendTrait for BaseBackend {
    impl_backend_trait! {}

    fn get_format(&self) -> Format {
        Format::Invalid
    }

    fn set_format(&mut self, _: Format) {
        // Nothing.
    }

    unsafe fn create_surface(&mut self, _: &mut Env) -> Option<Surface> {
        panic!("Cannot execute create_surface on BaseBackend!");
    }

    unsafe fn destroy_surface(&mut self, _: &mut Env) {
        if let Some(surface) = &mut self.surface {
            cairo_surface_destroy(surface.to_raw_none());
            self.surface = None;
        }
    }

    unsafe fn recreate_surface(&mut self, env: &mut Env) -> Option<Surface> {
        self.destroy_surface(env);
        self.create_surface(env)
    }

    unsafe fn set_canvas(&mut self, canvas: Canvas) {
        self.canvas = canvas;
    }
}
